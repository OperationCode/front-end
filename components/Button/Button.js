import React from 'react';
import { any, bool, func, node, number, oneOf, oneOfType, string } from 'prop-types';
import classNames from 'classnames';
import ReactGA from 'react-ga';
import { BUTTON } from 'common/constants/testIDs';
import { googleAnalyticsEventPropType } from 'common/constants/custom-props';
import { getDataAttributes, getAriaAttributes } from 'common/utils/prop-utils';
import styles from './Button.module.css';

Button.propTypes = {
  analyticsObject: googleAnalyticsEventPropType,
  children: oneOfType([node, string]).isRequired,
  className: string,
  disabled: bool,
  fullWidth: bool,
  onClick: func,
  tabIndex: oneOfType([string, number]),
  theme: oneOf(['primary', 'secondary']),
  type: oneOf(['button', 'reset', 'submit']),
  datum: any, // eslint-disable-line react/forbid-prop-types
};

Button.defaultProps = {
  analyticsObject: {
    action: 'Button Selected',
    category: 'Interactions',
  },
  className: undefined,
  datum: '',
  disabled: false,
  fullWidth: false,
  onClick: () => {},
  tabIndex: 0,
  theme: 'primary',
  type: 'button',
};

export default function Button({
  analyticsObject,
  children,
  className,
  disabled,
  fullWidth,
  onClick,
  tabIndex,
  theme,
  type,
  ...rest
}) {
  const customDataAttributes = getDataAttributes(rest);
  const ariaAttributes = getAriaAttributes(rest);

  const clickHandler = () => {
    if (process.env.NODE_ENV === 'production') {
      ReactGA.event(analyticsObject);
    } else {
      // eslint-disable-next-line no-console
      console.log('Analytics Disabled', analyticsObject);
    }
    onClick();
  };

  /* eslint-disable react/button-has-type */
  return (
    <button
      className={classNames(styles.Button, className, styles[theme], {
        [styles.disabled]: disabled,
        [styles.fullWidth]: fullWidth,
      })}
      data-testid={BUTTON}
      disabled={disabled}
      onClick={clickHandler}
      tabIndex={tabIndex}
      type={type}
      {...customDataAttributes}
      {...ariaAttributes}
    >
      {children}
    </button>
  );
  /* eslint-enable react/button-has-type */
}
