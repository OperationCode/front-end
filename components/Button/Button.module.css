.Button {
  display: inline-block;
  border: 3px solid transparent;
  border-radius: 3px;
  cursor: pointer;
  font-size: 1rem;
  font-weight: bold;
  text-align: center;
  text-decoration: none;
  text-transform: uppercase;
  vertical-align: middle;
  line-height: 1.25;
  min-width: 175px;
  padding: 0.75rem;
  white-space: nowrap;
  transition: 0.2s linear background-color, 0.2s linear color;
  text-shadow: none;
}

.primary {
  background-color: var(--primary);
  border-color: var(--primary);
  color: var(--secondary);
  outline-color: var(--primary);
}

.primary:hover:enabled,
.primary:focus:enabled {
  background-color: rgba(0, 0, 0, 0);
  color: var(--primary);
}

.secondary {
  background-color: var(--secondary);
  border-color: var(--secondary);
  color: var(--primary);
  outline-color: var(--secondary);
}

.secondary:hover:enabled,
.secondary:focus:enabled {
  background-color: rgba(0, 0, 0, 0);
  color: var(--secondary);
}

.fullWidth {
  width: 100%;
}

.disabled {
  border: 3px solid transparent;
  opacity: 0.6;
}

.disabled:hover {
  border: 3px solid transparent;
  cursor: not-allowed;
}

a.Button {
  font-family: var(--primaryFontFamily), sans-serif;
}
