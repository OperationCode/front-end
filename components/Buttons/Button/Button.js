import React from 'react';
import { any, bool, func, node, number, oneOf, oneOfType, string } from 'prop-types';
import noop from 'lodash/noop';
import classNames from 'classnames';
import { BUTTON } from 'common/constants/testIDs';
import { googleAnalyticsEventPropType } from 'common/constants/custom-props';
import { gtag } from 'common/utils/thirdParty/gtag';
import { getDataAttributes, getAriaAttributes } from 'common/utils/prop-utils';
import styles from './Button.module.css';

Button.propTypes = {
  analyticsObject: googleAnalyticsEventPropType,
  children: oneOfType([node, string]).isRequired,
  className: string,
  disabled: bool,
  fullWidth: bool,
  onClick: func, // why optional? form buttons typically act off of a form's submit/reset events
  tabIndex: oneOfType([string, number]),
  theme: oneOf(['primary', 'secondary']),
  type: oneOf(['button', 'reset', 'submit']),
  datum: any, // eslint-disable-line react/forbid-prop-types
};

Button.defaultProps = {
  analyticsObject: {
    action: 'Button Selected',
    category: 'Interactions',
  },
  className: undefined,
  datum: '',
  disabled: false,
  fullWidth: false,
  onClick: noop,
  tabIndex: 0,
  theme: 'primary',
  type: 'button',
};

export default function Button({
  analyticsObject,
  children,
  className,
  disabled,
  fullWidth,
  onClick,
  tabIndex,
  theme,
  type,
  ...rest
}) {
  const customDataAttributes = getDataAttributes(rest);
  const ariaAttributes = getAriaAttributes(rest);

  const eventConfig = {
    ...analyticsObject,
    label: typeof children === 'string' ? children : undefined,
  };

  const clickHandler = () => {
    gtag.event(eventConfig);
    onClick();
  };

  /* eslint-disable react/button-has-type */
  return (
    <button
      className={classNames(styles.Button, className, styles[theme], {
        [styles.disabled]: disabled,
        [styles.fullWidth]: fullWidth,
      })}
      data-testid={BUTTON}
      disabled={disabled}
      onClick={clickHandler}
      tabIndex={tabIndex}
      type={type}
      {...customDataAttributes}
      {...ariaAttributes}
    >
      {children}
    </button>
  );
  /* eslint-enable react/button-has-type */
}
