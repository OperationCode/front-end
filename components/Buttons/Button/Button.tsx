import React from 'react';
import noop from 'lodash/noop';
import classNames from 'classnames';
import { BUTTON } from 'common/constants/testIDs';
import { gtag } from 'common/utils/thirdParty/gtag';
import { getDataAttributes, getAriaAttributes } from 'common/utils/prop-utils';
import styles from './Button.module.css';

type GoogleAnalyticsEventPropType = {
  /**
   * A description of the behaviour. E.g. 'Clicked Delete', 'Added a component', 'Deleted account'
   */
  action: string;
  /**
   * A top level category for these events. E.g. 'User', 'Navigation', 'App Editing', etc.
   */
  category: string;
  /**
   * More precise labeling of the related action. E.g. alongside the 'Added a component' action,
   * we could add the name of a component as the label. E.g. 'Survey', 'Heading', 'Button', etc.
   */
  label?: string;
  /**
   * A means of recording a numerical value against an event. E.g. a rating, a score, etc.
   */
  value?: number;
  /**
   * If an event is not triggered by a user interaction, but instead by our code (e.g. on page
   * load), it should be flagged as a nonInteraction event to avoid skewing bounce rate data.
   */
  nonInteraction?: boolean;
  /**
   * This specifies the transport mechanism with which hits will be sent. Valid values include
   * 'beacon', 'xhr', or 'image'.
   */
  transport?: 'beacon' | 'xhr' | 'image';
};

type ButtonProps = {
  /**
   * Helps track in-page `event` interactions.
   */
  analyticsObject?: GoogleAnalyticsEventPropType;
  /**
   * Forces the component's width as wide as its parent container's width.
   */
  fullWidth?: boolean;
  /**
   * Sets the button color theme.
   */
  theme?: 'primary' | 'secondary';
} & React.ButtonHTMLAttributes<HTMLButtonElement>;

export default function Button({
  analyticsObject = {
    action: 'Button Selected',
    category: 'Interactions',
  },
  children,
  className = undefined,
  disabled = false,
  fullWidth = false,
  onClick = noop,
  tabIndex = 0,
  theme = 'primary',
  type = 'button',
  ...rest
}: ButtonProps) {
  const customDataAttributes = getDataAttributes(rest);
  const ariaAttributes = getAriaAttributes(rest);

  const eventConfig = {
    ...analyticsObject,
    label: typeof children === 'string' ? children : undefined,
  };

  return (
    <button
      className={classNames(styles.Button, className, styles[theme], {
        [styles.disabled]: disabled,
        [styles.fullWidth]: fullWidth,
      })}
      data-testid={BUTTON}
      disabled={disabled}
      onClick={e => {
        gtag.event(eventConfig);
        onClick(e);
      }}
      tabIndex={tabIndex}
      type={type}
      {...customDataAttributes}
      {...ariaAttributes}
    >
      {children}
    </button>
  );
}
