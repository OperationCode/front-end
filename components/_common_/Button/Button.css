.Button {
  align-items: center;
  border: 1px solid transparent;
  border-radius: 4px;
  color: #ffffff;
  cursor: pointer;
  display: flex;
  font-size: 1rem;
  justify-content: center;
  line-height: 1.25;
  min-width: 200px;
  padding: 0;
  transition: 0.2s linear background-color;
  text-decoration: none;
}

.Button:active,
.Button:focus {
  outline-color: #000000;
  outline-offset: -2px;
  outline-style: solid;
  outline-width: 2px;
}

.Button > * {
  /* Apply padding to child so that when child is an anchor element width and height span the
  clickable area of the button */
  padding: 0.75rem;
}

.Button > a {
  /* Hijack anchor styles if nested within a Button */
  color: #ffffff;
  height: 100%;
  text-decoration: none;
  width: 100%;
}

.disabled {
  border: none;
  opacity: 0.6;
}

.disabled:hover {
  border: none;
  cursor: not-allowed;
}

.fullWidth {
  width: 100%;
}

.primary {
  background-color: var(--primary);
  border-color: var(--primaryDark);
  outline-color: var(--primaryDark);
}

.primary:hover:enabled {
  background-color: var(--primaryLight);
}

.primary:active:not:hover {
  background-color: var(--primaryDark);
}

.secondary {
  background-color: var(--secondary);
  border-color: var(--secondaryDark);
  outline-color: var(--secondaryDark);
}

.secondary:hover:enabled {
  background-color: var(--secondaryLight);
}

.secondary:active:not:hover {
  background-color: var(--secondaryDark);
}

.slate {
  background-color: var(--slate);
  border-color: var(--slateDark);
  outline-color: var(--slateDark);
}

.slate:hover:enabled {
  background-color: var(--slateLight);
}

.slate:active:not:hover {
  background-color: var(--slateDark);
}
