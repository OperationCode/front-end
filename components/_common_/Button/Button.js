import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import ReactGA from 'react-ga';
import { googleAnalyticsEventPropType } from 'common/constants/custom-props';
import styles from './Button.css';

class Button extends Component {
  static propTypes = {
    analyticsObject: googleAnalyticsEventPropType,
    children: PropTypes.oneOfType([PropTypes.node, PropTypes.string]).isRequired,
    className: PropTypes.string,
    disabled: PropTypes.bool,
    fullWidth: PropTypes.bool,
    onClick: PropTypes.func,
    tabIndex: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    theme: PropTypes.oneOf(['primary', 'secondary', 'slate']),
    type: PropTypes.oneOf(['button', 'reset', 'submit']),
  };

  static defaultProps = {
    analyticsObject: {
      category: 'Interactions',
      action: 'Button Selected',
    },
    className: '',
    disabled: false,
    fullWidth: false,
    onClick: undefined,
    tabIndex: 0,
    theme: 'primary',
    type: 'button',
  };

  clickHandler = () => {
    const { props } = this;

    if (process.env.NODE_ENV === 'production') {
      ReactGA.event(props.analyticsObject);
    } else {
      // eslint-disable-next-line no-console
      console.log('Analytics Disabled', props.analyticsObject);
    }

    return props.onClick;
  };

  render() {
    const { props } = this;

    /* eslint-disable react/button-has-type */
    return (
      <button
        className={classNames(styles.Button, props.className, styles[props.theme], {
          [styles.disabled]: props.disabled,
          [styles.fullWidth]: props.fullWidth,
        })}
        disabled={props.disabled}
        onClick={this.clickHandler}
        tabIndex={props.tabIndex}
        type={props.type}
      >
        {/* Render text nodes within a span to apply selector styles */}
        {typeof props.children === 'string' ? <span>{props.children}</span> : props.children}
      </button>
    );
    /* eslint-enable react/button-has-type */
  }
}

export default Button;
