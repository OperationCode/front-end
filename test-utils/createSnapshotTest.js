/* eslint-env jest */
import renderer from 'react-test-renderer';

/**
 * Used to create fully-rendered snapshot test.
 * Useful for reusable component smoke tests.
 *
 * @export
 * @param {*} Component
 */
export default Component => {
  const tree = renderer.create(Component).toJSON();
  expect(tree).toMatchSnapshot();
};
