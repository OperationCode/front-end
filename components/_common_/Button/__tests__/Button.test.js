/* eslint-env jest */
import React from 'react';
import { mount, shallow } from 'enzyme';
import createSnapshotTest from 'test-utils/createSnapshotTest';

import Button from '../Button';

describe('Button', () => {
  test('should render with just required props passed', () => {
    createSnapshotTest(<Button>Test</Button>);
  });

  test('should render properly with some props assigned', () => {
    createSnapshotTest(
      <Button
        theme="secondary"
        data-custom-attr="custom stuff here"
        disabled
        fullWidth
        type="submit"
      >
        Test
      </Button>,
    );
  });

  test('should render without a generated span when children is PropTypes.node', () => {
    const testText = 'Testing No Span';

    const ButtonInstance = mount(
      <Button>
        <b>{testText}</b>
      </Button>,
    );

    expect(
      ButtonInstance.containsAnyMatchingElements([
        <span>{testText}</span>,
        <span>
          <b>{testText}</b>
        </span>,
      ]),
    ).toEqual(false);
  });

  test('should render with a generated span when children is PropTypes.string', () => {
    const testText = 'Testing With Span';

    const ButtonInstance = mount(<Button>{testText}</Button>);

    expect(ButtonInstance.containsAllMatchingElements([<span>{testText}</span>])).toEqual(true);
  });

  test('should send log to console when clickHandler is called in non-prod environment', () => {
    /* eslint-disable no-console */
    console.log = jest.fn();

    const ButtonShallowInstance = shallow(<Button>Testing</Button>);

    ButtonShallowInstance.instance().clickHandler();

    expect(console.log.mock.calls.length).toEqual(1);
    /* eslint-enable no-console */
  });
});
