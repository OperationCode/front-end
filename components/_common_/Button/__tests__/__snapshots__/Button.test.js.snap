// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should render properly with some props assigned 1`] = `
<button
  className="Button secondary disabled fullWidth"
  disabled={true}
  onClick={[Function]}
  tabIndex={0}
  type="submit"
>
  <span>
    Test
  </span>
</button>
`;

exports[`Button should render with just required props passed 1`] = `
<button
  className="Button primary"
  disabled={false}
  onClick={[Function]}
  tabIndex={0}
  type="button"
>
  <span>
    Test
  </span>
</button>
`;
